witchcraft/internal/tcpjson/conn_provider.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpjson
+
+import (
+	"crypto/tls"
+	"net"
+	"net/url"
+	"sync/atomic"
+
+	werror "github.com/palantir/witchcraft-go-error"
+)
+
+// ConnProvider defines the behavior to retrieve an established net.Conn.
+type ConnProvider interface {
+	// GetConn returns a net.Conn or an error if there is no connection established.
+	// It is the users responsibility to close the returned net.Conn.
+	GetConn() (net.Conn, error)
+}
+
+const (
+	ErrNoURIs           = "no URIs provided to connect to"
+	ErrFailedParsingURI = "failed to parse provided URI"
+	ErrFailedDial       = "failed to dial the host:port"
+)
+
+var _ ConnProvider = (*TCPConnProvider)(nil)
+
+// TCPConnProvider implements a ConnProvider that will round-robin TCP connections to a specific set of hosts.
+type TCPConnProvider struct {
```

Comment:
I usually like to return concrete types in situations where an implementation could satisfy multiple interfaces rather than defining a single large interface that groups everything. In this case there is only 1 method, so I am fine to do make it private and return the interface.